userspace/async/async_metadata_source.h

```diff
@@ -0,0 +1,237 @@
+	 *                    are available.  If this handler is empty, then
+	 *                    this async_metadata_source will store the metadata
+	 *                    and return them on the next call to lookup().
+	 *
+	 * @returns true if this method was able to lookup and return the
+	 *          metadata synchronously; false otherwise.
+	 */
+	bool lookup(const key_type& key,
+                    metadata_type& metadata,
+                    const callback_handler& handler = callback_handler());
+
+	/**
+	 * Determines if the async thread assocaited with this
+	 * async_metadata_source is running.
+	 *
+	 * <b>Note:</b> This API is for information only.  Clients should
+	 * not use this to implement any sort of complex behavior.  Such
+	 * use will lead to race conditions.  For example, is_running() and
+	 * lookup() could potentially race, causing is_running() to return
+	 * false after lookup() has started the thread.
+	 *
+	 * @returns true if the async thread is running, false otherwise.
+	 */
+	bool is_running() const;
+
+protected:
+	/**
+	 * Stops the thread assocaited with this async_metadata_source, if
+	 * it is running.
+	 */
+	void stop();
+
+	/**
+	 * Returns the number of elements in the requeust queue.  Concrete
+	 * subclasses will call this methods from their run_impl() methods to
+	 * determine if there is more asynchronous work from them to perform.
+	 *
+	 * @returns the size of the request queue.
+	 */
+	std::size_t queue_size() const;
+
+	/**
+	 * Dequeues an entry from the request queue and returns it.  Concrete
+	 * subclasses will call this method to get the next key for which
+	 * to collect metadata.
+	 *
+	 * Precondition: queue_size() must be non-zero.
+	 *
+	 * @returns the next key to look up.
+	 */
+	key_type dequeue_next_key();
+
+	metadata_type get_metadata(const key_type& key);
+
+	/**
+	 * Stores a collected set of metadata for the given key.  Concrete
+	 * subclasses will call this method from their run_impl() method to
+	 * save (or otherwise notifiy the client about) a collected metadata.
+	 *
+	 * @param[in] key      The key for which the client asked for metadata.
+	 * @param[in] metadata The collected metadata.
+	 */
+	void store_metadata(const key_type& key, const metadata_type& metadata);
+
+	/**
+	 * Concrete subclasses must override this method to perform the
+	 * asynchronous metadata lookup.
+	 */
+	virtual void run_impl() = 0;
+
+private:
+	struct lookup_request
+	{
+		lookup_request():
+			m_available(false),
+			m_metadata(),
+			m_available_condition(),
+			m_callback(),
+			m_start_time(std::chrono::steady_clock::now())
+		{ }
+
+		bool m_available;
+		metadata_type m_metadata;
+		std::condition_variable m_available_condition;
+		callback_handler m_callback; // TODO: This may need to be a list
+		std::chrono::time_point<std::chrono::steady_clock> m_start_time;
+	};
+
+	typedef std::map<key_type, lookup_request> metadata_map;
+
+	void run();
```

Comment:
as per above, we need some comments. It's not inherently clear where functions like this fit in the call chain